tech_reports/LLMs/llms.md

```diff
@@ -55,14 +55,294 @@ Authors:
 ### 2.7 LM Head
 ## 3. Features
 ### 3.1 Generative Decoding
+
 ### 3.2 Prefill and Decode
-  - submodules, tests
-  - how to combine prefill and decode, 
-  - slicing prefill to fit in L1
+
+Large language models require two distinct phases for inference due to the fundamental nature of transformer attention and autoregressive generation: prefill and decode.
+
+The prefill phase is done sequentially for each user, but parallel for the prompt tokens of each user. During prefill, the model computes attention scores for all prompt tokens against each other and populates the key-value (KV) cache which will speed up the computation of the decode phase. At the end of the prefill phase, the first token for the following autoregressive generation will also be computed.
+
+The decode phase is parallel-computed for all users, but sequential for each token within a batch of users. Each new token can only be generated after the previous one, as the model must maintain causality in attention computations.
+
+#### Llama3 Module and Test Differences
+
+In our current Llama3 model, the attention module class (`TtLlamaAttention`) implements two primary methods for attention computation: `forward_prefill` and `forward_decode`. 
+To test these, we provide two separate attention test files, `test_attention_decode` and `test_attention_prefill`, which create the appropriate input tensors:
+- A tensor of size `(batch, dim)` in L1 for decode,
+- A tensor of size `(seqlen, dim)` in DRAM for prefill.
+
+Each attention test compares the attention module output and KV-cache correlation between the PyTorch host implementation and the TTNN device implementation.
+
+The current version of the MLP module class (`TtLlamaMLP`) handles prefill and decode in the same file but has some technical differences (mentioned in the section below).
+
+The decoder module (which encapsulates both attention and MLP) and model module (which encapsulates the decoder and the remaining parts of the Llama3 model) also handle prefill and decode in the same file, but they call the respective modes within the attention and MLP modules.
```

Comment:
I think this part is a bit confusing. I understand what we want to convey but it might be best to just cut if out.
It doesn't add much to the overall section. Thoughts?

Reply:
yeah I just added it because there was a bullet point about modules/tests in the outline - prefer to remove

Reply:
I think it can stay, people might find it useful. However, it is a little out of place here, maybe it is better if it comes after implementation details section.